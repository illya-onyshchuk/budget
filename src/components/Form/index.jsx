import { Component } from 'react';
import PropTypes from 'prop-types';

class Form extends Component {
  constructor() {
    super();

    this.state = {
      value: '',
    }
  }

  onSubmit = (e) => {
    e.preventDefault()

    this.props.onChange(this.state.value)
    this.setState({value: ''} )
  }

  onChange = (e) => {
    const {value} = e.target;
  
    this.setState({
      value: +value
    })
  }
 
  render() {
    return ( 
      <form onSubmit={this.onSubmit}>
        <input
        type="number" 
        name='balance' 
        placeholder='Sum'
        value={this.state.value}
        onChange={this.onChange}
        />
        <button>Seve</button>
      </form>
    );
  }
}

Form.propTypes = {
  onChange: PropTypes.func
}
 
export default Form;