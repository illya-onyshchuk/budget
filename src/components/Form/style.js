import { styled } from "styled-components";

export const Wrapper = styled.div`
  background: #fff;
  border: 1px solid #dbdbdb;
  border-radius: 3px;
  padding: 5px;
  margin-bottom: 10px;
`;
Wrapper.displayName = 'TransactionWrapper'


export const Input = styled.input`
  border-radius: 8px;
  padding: 0 5px;
  min-width: 80px;
  margin: 10px 0;
  margin-right: 10px;
  border: 1px solid #dbdbdb;
  height: 40px;
  outline: none;
  box-shadow: 7px 7px 14px #b6a9a9, 7px -7px 14px #f6e5e5;
`;
Input.displayName = 'Input'


export const Row = styled.div`
  display: flex;
  align-items: center;
`;
Row.displayName = 'Row'

export const Comment = styled.textarea`
  padding: 5px;
  margin-left: 10px;
  border-radius: 8px;
  resize: none;
  color: black;
  height: 40px;
  border: 1px solid #b9acac;
  background-color: #e0e0e0;
  font-family: ingerit;
  line-height: 1.5;
  min-width: 200px;
  box-shadow: 10px 7px 14px #b6a9a9, 7px -7px 14px #f6e5e5;
`;
Comment.displayName = 'Comment';

export const Button = styled.button`
    background-color: #e0e0e0;
    height: 40px;
    padding: 0 21px;
    margin-right: 5px;
    text-transform: uppercase;
    letter-spacing: 2px;
    cursor: pointer;
    border-radius: 10px;
    border: 2px solid #e0e0e0;
    // box-shadow: rgba(50, 50, 93, 0.25) 0px 2px 5px -1px, rgba(0, 0, 0, 0.3) 0px 1px 3px -1px;
    box-shadow: 7px 7px 14px #b6a9a9, 7px -7px 14px #f6e5e5;
    transition: .4s;
  
  
   & span:last-child {
    display: none;
  }
  
  &:hover {
    transition: .4s;
    border: 2px dashed #00BFA6;
    background-color: #fff;
    color: #00BFA6;
  }
  
  &:active {
    background-color: #87dbd0;
  } 
`;
Button.displayName = 'Button';
