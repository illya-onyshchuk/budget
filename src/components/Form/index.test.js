import React  from "react";
import Form from '.';
import { shallow } from 'enzyme';

describe('Form component', () => {
  let props;
  let sut;

  beforeEach(() => {
      jest
        .spyOn(global, 'Date')
        .mockImplementation(() => ({
          toISOString: () => '2023-01-01T16:17:20.220Z'
        }))
  })

  beforeEach(() => {
    props = {
      onChange: jest.fn()
    }

    sut = shallow(<Form {...props}/>)
  });

  describe('when edit form', () => {
    it('should change value input', () => {
      let input = sut.find('Input[name="value"]').at(0);

      input.simulate('change', {target: {
        value: '123',
        name: 'value'
      }});

      sut.update();
      input = sut.find('Input[name="value"]').at(0);

      const props = input.props();
      expect(props.value).toBe("123");
    });

    it('should change comment input', () => {
      const expectedResult = 'comment value';
      let input = sut.find('Comment[name="comment"]').at(0);

      input.simulate('change', {target: {
        value: 'comment value',
        name: 'comment'
      }});

      sut.update();
      input = sut.find('Comment[name="comment"]').at(0);

      const {value} = input.props();
      expect(value).toBe(expectedResult);
    });
  });

  describe('when user submit form', () => {
    it('should call onChage form the props', () => {
      const form = sut.find('form');

      form.simulate('submit', {
        preventDefault: jest.fn()
      });

      expect(props.onChange).toHaveBeenCalledTimes(1);

    });

    it('should send form date', () => {
      let input = sut.find('Input[name="value"]').at(0);

      input.simulate('change', {target: {
        value: '123',
        name: 'value'
      }});

      input = sut.find('Comment[name="comment"]').at(0);

      input.simulate('change', {target: {
        value: 'comment value',
        name: 'comment'
      }});

      sut.update();

      const form = sut.find('form');
      form.simulate('submit', {
        preventDefault: jest.fn()
      });

      expect(props.onChange).toHaveBeenCalledWith({
        value: '123',
        date: '2023-01-01',
        comment: 'comment value',
      })
    })
  })
});