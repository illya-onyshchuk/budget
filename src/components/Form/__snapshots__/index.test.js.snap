// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transaction component should show transaction 1`] = `
<form
  onSubmit={[Function]}
>
  <Row>
    <Input
      name="date"
      onChange={[Function]}
      type="date"
      value="2023-08-01"
    />
    <Input
      name="value"
      onChange={[Function]}
      placeholder="Sum"
      type="number"
      value=""
    />
  </Row>
  <Row>
    <Button>
      Save
    </Button>
    <Comment
      name="comment"
      onChange={[Function]}
      value=""
    />
  </Row>
</form>
`;
